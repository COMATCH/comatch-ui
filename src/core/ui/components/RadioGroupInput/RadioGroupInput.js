import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import get from 'lodash/get';
import classNames from 'classnames';
import { InputLabel } from '../InputLabel'
import { InputError } from '../InputError'
import { RadioInput } from '../RadioInput'

import './RadioGroupInput.scss';

const propTypes = {
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
    className: PropTypes.string,
    disabled: PropTypes.bool,
    display: PropTypes.oneOf(['block', 'inline']),
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    options: PropTypes.arrayOf(PropTypes.shape({
        disabled: PropTypes.bool,
        label: PropTypes.string,
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
    })).isRequired,
    required: PropTypes.bool,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
};

const defaultProps = {
    children: null,
    className: null,
    disabled: false,
    display: 'block',
    id: null,
    inputError: '',
    label: '',
    onChange: noop,
    required: false,
    value: null,
};

export const RadioGroupInput = ({
    className,
    children,
    disabled,
    display,
    id,
    inputError,
    label,
    name,
    onChange,
    options,
    required,
    value,
}) => {
    const classes = classNames('Input', 'RadioGroupInput', { inline: display === 'inline' }, className);

    const handleRadioChange = (event) => {
        if (event.target.checked) {
            onChange(event);
        }
    };

    return (
        <div id={id} className={classes}>
            {label && <InputLabel text={label} required={required} />}
            {options.map((option, index) => (
                <RadioInput
                    key={index}
                    display={display}
                    label={option.label}
                    disabled={disabled || option.disabled}
                    name={name}
                    onChange={handleRadioChange}
                    value={option.value}
                    checked={option.value === value}
                />
            ))}
            {React.Children.map(children, (child) => {
                if (child.type === RadioInput) {
                    return React.cloneElement(child, {
                        onChange: handleRadioChange,
                        checked: get(child, 'props.value') === value,
                    });
                }
                return child;
            })}
            {inputError && <InputError text={inputError} />}
        </div>
    );
};

RadioGroupInput.defaultProps = defaultProps;

RadioGroupInput.propTypes = propTypes;
