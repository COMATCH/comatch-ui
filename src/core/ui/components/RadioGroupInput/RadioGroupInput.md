## RadioGroupInput

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**children** | `Union<Array[]<ReactNode>\|ReactNode>` | `null` | :x: | 
**children<1>** | `Array[]<ReactNode>` |  | :x: | 
**children<2>** | `ReactNode` |  | :x: | 
**className** | `String` | `null` | :x: | 
**disabled** | `Boolean` | `false` | :x: | 
**display** | `Enum('block', 'inline')` | `'block'` | :x: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `Union<String\|ReactNode>` | `''` | :x: | 
**label<1>** | `String` |  | :x: | 
**label<2>** | `ReactNode` |  | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onChange** | `Function` | `noop` | :x: | 
**options** | `Array[]<Shape>` |  | :white_check_mark: | 
**options[].disabled** | `Boolean` |  | :x: | 
**options[].label** | `String` |  | :x: | 
**options[].value** | `Union<String\|Number\|Boolean>` |  | :x: | 
**options[].value<1>** | `String` |  | :x: | 
**options[].value<2>** | `Number` |  | :x: | 
**options[].value<3>** | `Boolean` |  | :x: | 
**required** | `Boolean` | `false` | :x: | 
**value** | `Union<String\|Number\|Boolean>` | `null` | :x: | 
**value<1>** | `String` |  | :x: | 
**value<2>** | `Number` |  | :x: | 
**value<3>** | `Boolean` |  | :x: | 

