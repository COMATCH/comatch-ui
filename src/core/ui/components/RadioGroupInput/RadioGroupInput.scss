@import '../../../styles/variables.scss';

.RadioGroupInput {
    margin-bottom: $co-inputs-margin-bottom;

    span.InputLabel {
        // Whitespace
        & ~ .RadioInput:not(.block) {
            margin-left: 10px;
        }
    }
}
