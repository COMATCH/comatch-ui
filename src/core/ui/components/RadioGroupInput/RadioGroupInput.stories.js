import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { RadioGroupInput } from './RadioGroupInput';

const options = [
    { label: 'Apple', value: 'apple' },
    { label: 'Banana', value: 'banana' },
    { label: 'Peach', value: 'peach' },
];

storiesOf('RadioGroupInput', module)
    .add('basic use', () => <RadioGroupInputContainer name="fruit" options={options} />)
    .add('with label', () => <RadioGroupInputContainer name="fruit" options={options} label="Pick a fruit" />)
    .add('with display inline', () => <RadioGroupInputContainer name="fruits" options={options} display="inline" />)
    .add('with error', () => (
        <RadioGroupInputContainer name="fruits" options={options} inputError="I am the ultimate error" />
    ));

class RadioGroupInputContainer extends Component {
    state = { fruit: 'banana' };

    handleChange = (event) => {
        this.setState({ fruit: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };

    render() {
        return <RadioGroupInput onChange={this.handleChange} value={this.state.fruit} {...this.props} />;
    }
}
