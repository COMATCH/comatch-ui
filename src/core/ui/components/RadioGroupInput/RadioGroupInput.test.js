import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import noop from 'lodash/noop';
import { RadioGroupInput } from './RadioGroupInput';
import '../../../config/tests/setup';

describe('RadioGroupInput tests', () => {
    const REQUIRED_PROPS = {
        name: 'some name',
        options: [
            { label: 'Apple', value: 'apple' },
            { label: 'Banana', value: 'banana' },
            { label: 'Peach', value: 'peach' },
        ],
    };

    const OPTIONAL_PROPS = {
        className: 'RadioInput block',
        id: null,
        checked: false,
        disabled: false,
        onChange: noop,
        type: 'radio',
        value: 'apple',
    };

    it('should render and assert snapshot', () => {
        const radioGroupInputComponent = renderer.create(<RadioGroupInput {...REQUIRED_PROPS} />).toJSON();
        expect(radioGroupInputComponent).toMatchSnapshot();
    });

    it('should render correctly with optional props', () => {
        const props = {
            ...OPTIONAL_PROPS,
            ...REQUIRED_PROPS,
        };
        const radioGroupInputComponent = mount(<RadioGroupInput {...props} />);
        expect(radioGroupInputComponent.prop('className')).toEqual(props.className);
        expect(radioGroupInputComponent.prop('id')).toEqual(props.id);
        expect(radioGroupInputComponent.prop('checked')).toEqual(props.checked);
        expect(radioGroupInputComponent.prop('disabled')).toEqual(props.disabled);
        expect(radioGroupInputComponent.prop('type')).toEqual(props.type);
        expect(radioGroupInputComponent.prop('value')).toEqual(props.value);
        expect(radioGroupInputComponent.prop('onChange')).toEqual(props.onChange);
    });

    it('should handle the onChange correctly', () => {
        const props = {
            ...OPTIONAL_PROPS,
            ...REQUIRED_PROPS,
        };
        const mockFunction = jest.fn();
        const wrapper = mount(<RadioGroupInput {...props} onChange={mockFunction} />);

        wrapper.find('input').at(0).simulate('change', { target: { checked: true } });
        wrapper.find('input').at(1).simulate('change', { target: { checked: true } });
        expect(mockFunction).toHaveBeenCalledTimes(2);
    });

    describe('verifying the HTML structure', () => {
        const props = {
            ...REQUIRED_PROPS,
        };
        it('should expect nodes/dom elements', () => {
            const radioGroupInputComponent = mount(<RadioGroupInput {...props} />);
            expect(radioGroupInputComponent).toHaveLength(1);
            expect(radioGroupInputComponent.find('div')).toHaveLength(7);
            expect(radioGroupInputComponent.find('div').at(0).hasClass('Input RadioGroupInput')).toBe(true);

            expect(radioGroupInputComponent.find('div').find('div')).toHaveLength(7);
            expect(radioGroupInputComponent.find('div').find('div').at(1).hasClass('RadioInput block')).toBe(true);
            expect(radioGroupInputComponent.find('div').find('div').find('label')).toHaveLength(3);
            expect(radioGroupInputComponent.find('div').find('div').find('label').find('input')).toHaveLength(3);

            const listItems = radioGroupInputComponent.find('div').find('div').find('label');
            expect(listItems.find('input').at(0).props().value).toEqual('apple');
            expect(listItems.find('input').at(0).props().name).toEqual('some name');

            expect(listItems.find('div')).toHaveLength(3);
            expect(listItems.find('div').at(0).hasClass('RadioInput__input')).toBe(true);
            expect(listItems.find('div').at(0).props().role).toEqual('radio');

            expect(listItems.find('span')).toHaveLength(3);
            expect(listItems.find('span').at(0).hasClass('InputLabel RadioInput__label')).toBe(true);
            expect(listItems.find('span').at(0).text()).toEqual('Apple');
        });

        it('should expect nodes/dom elements with `label`', () => {
            const radioGroupInputComponent = mount(<RadioGroupInput {...props} label={'Input label text'}/>);
            expect(radioGroupInputComponent.find('div').find('label').at(0).hasClass('InputLabel')).toBe(true);
            expect(radioGroupInputComponent.find('div').find('label').at(0).text()).toEqual('Input label text');
        });

        it('should expect nodes/dom elements with display `inline`', () => {
            const radioGroupInputComponent = mount(<RadioGroupInput {...props} display={'inline'}/>);
            expect(radioGroupInputComponent.find('div').at(0).hasClass('Input RadioGroupInput inline')).toBe(true);
            expect(radioGroupInputComponent.find('div').find('div').at(1).hasClass('RadioInput inline')).toBe(true);
        });

        it('should expect nodes/dom elements with `error label`', () => {
            const radioGroupInputComponent = mount(<RadioGroupInput {...props} inputError={'error message'}/>);
            expect(radioGroupInputComponent.find('div').find('span').at(3).hasClass('InputError')).toBe(true);
            expect(radioGroupInputComponent.find('div').find('span').at(3).text()).toEqual('error message');
        });
    });
});
