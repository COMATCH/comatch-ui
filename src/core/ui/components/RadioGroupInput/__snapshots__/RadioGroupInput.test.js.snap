// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroupInput tests should render and assert snapshot 1`] = `
<div
  className="Input RadioGroupInput"
  id={null}
>
  
  <div
    className="RadioInput block"
    id={null}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        id={null}
        name="some name"
        onChange={[Function]}
        type="radio"
        value="apple"
      />
      <div
        className="RadioInput__input"
        role="radio"
      />
      <span
        className="sc-bdVaJa eCYGDq InputLabel RadioInput__label"
        id={null}
      >
        Apple
      </span>
    </label>
    
  </div>
  <div
    className="RadioInput block"
    id={null}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        id={null}
        name="some name"
        onChange={[Function]}
        type="radio"
        value="banana"
      />
      <div
        className="RadioInput__input"
        role="radio"
      />
      <span
        className="sc-bdVaJa eCYGDq InputLabel RadioInput__label"
        id={null}
      >
        Banana
      </span>
    </label>
    
  </div>
  <div
    className="RadioInput block"
    id={null}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        id={null}
        name="some name"
        onChange={[Function]}
        type="radio"
        value="peach"
      />
      <div
        className="RadioInput__input"
        role="radio"
      />
      <span
        className="sc-bdVaJa eCYGDq InputLabel RadioInput__label"
        id={null}
      >
        Peach
      </span>
    </label>
    
  </div>
  
</div>
`;
